carma_cooperative_perception/src/multiple_object_tracker_component.cpp

```diff
@@ -452,6 +453,103 @@ static auto predict_track_states(std::vector<Track> tracks, units::time::second_
+
+      // Fall back to 2D Euclidean distance if either semantic class if unknown. The unknown
+      // track/detection may actually be the same other track/detection we are scoring against.
+      if (
+        track.semantic_class == mot::SemanticClass::kUnknown ||
+        detection.semantic_class == mot::SemanticClass::kUnknown) {
+        return dist;
+      }
+
+      if (track.semantic_class == detection.semantic_class) {
+        return dist;
+      }
+    }
+
+    return std::nullopt;
+  }
+
+  template <typename... TrackAlternatives, typename... DetectionAlternatives>
+  auto operator()(
+    const std::variant<TrackAlternatives...> & track,
+    const std::variant<DetectionAlternatives...> & detection) const -> std::optional<float>
+  {
+    return std::visit(
+      [this](const auto & t, const auto & d) { return (*this)(t, d); }, track, detection);
+  }
+
+private:
+  static auto two_dimensional_distance(const mot::CtrvState & lhs, const mot::CtrvState & rhs)
+    -> float
+  {
+    const auto x_diff_sq{
+      std::pow(mot::remove_units(lhs.position_x) - mot::remove_units(rhs.position_x), 2)};
+    const auto y_diff_sq{
+      std::pow(mot::remove_units(lhs.position_y) - mot::remove_units(rhs.position_y), 2)};
+
+    return std::sqrt(x_diff_sq + y_diff_sq);
+  }
+
+  static auto two_dimensional_distance(const mot::CtraState & lhs, const mot::CtraState & rhs)
+    -> float
+  {
+    const auto x_diff_sq{
+      std::pow(mot::remove_units(lhs.position_x) - mot::remove_units(rhs.position_x), 2)};
+    const auto y_diff_sq{
+      std::pow(mot::remove_units(lhs.position_y) - mot::remove_units(rhs.position_y), 2)};
+
+    return std::sqrt(x_diff_sq + y_diff_sq);
+  }
+};
+
+/**
+ * @brief Calculates distance between a point and detection in SE(2) (special Euclidean) space
+*/
+struct MetricSe2
+{
+  template <typename Detection>
+  auto operator()(const mot::Point & point, const Detection & detection) const -> double
+  {
+    double sum{0};
+
+    sum += std::pow(mot::remove_units(point.position_x - detection.state.position_x), 2);
+    sum += std::pow(mot::remove_units(point.position_y - detection.state.position_y), 2);
+
+    const auto p_yaw_rad{mot::remove_units(point.yaw.get_angle())};
+    const auto d_yaw_rad{mot::remove_units(detection.state.yaw.get_angle())};
+    const auto abs_diff{std::abs(p_yaw_rad - d_yaw_rad)};
+    sum += std::min(abs_diff, 2 * 3.14159265359 - abs_diff);
+
+    return sum;
+  }
+
+  template <typename... Alternatives>
+  auto operator()(mot::Point point, const std::variant<Alternatives...> & detection) const -> double
+  {
+    const mot::Visitor visitor{
+      [this](const mot::Point & p, const mot::CtrvDetection & d) { return this->operator()(p, d); },
+      [this](const mot::Point & p, const mot::CtraDetection & d) { return this->operator()(p, d); },
+      [](const mot::Point &, const auto &) { return std::numeric_limits<double>::max(); }};
```

Comment:
Should we add a note here on why we are returning double max. Essentially because the passed types are not supported  by the MetricSe2